pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -358,9 +358,12 @@ func ResolvePipelineRunTask(
 		PipelineTask: &task,
 	}
 
+	isTaskRefCustomTask := rprt.PipelineTask.TaskRef != nil && rprt.PipelineTask.TaskRef.APIVersion != "" &&
+		rprt.PipelineTask.TaskRef.Kind != ""
+	isTaskSpecCustomTask := rprt.PipelineTask.TaskSpec != nil && rprt.PipelineTask.TaskSpec.APIVersion != "" &&
+		rprt.PipelineTask.TaskSpec.Kind != ""
 	cfg := config.FromContextOrDefaults(ctx)
-	rprt.CustomTask = cfg.FeatureFlags.EnableCustomTasks && rprt.PipelineTask.TaskRef != nil &&
-		rprt.PipelineTask.TaskRef.APIVersion != "" && rprt.PipelineTask.TaskRef.Kind != ""
+	rprt.CustomTask = cfg.FeatureFlags.EnableCustomTasks && (isTaskRefCustomTask || isTaskSpecCustomTask)
```

Comment:
Please separate this entire logic into a separate function to check whether a task is a custom task and assign `rprt.CustomTask` appropriate value here based on that new function. `ResolvePipelineRunTask` will be simplified instead of making everything part of it.